### ISSUE TITLE
Installation of documentation files and `manifest` files for dependencies

### ISSUE THREAD

**Issue by lyrahgames** [build2/build2#317]
When installing a package, the documentation files of its dependencies are not installed. To reproduce this, run the following commands.
```
git clone https://git.build2.org/hello/hello.git
bdep init -C @gcc cc config.cxx=g++ config.install.root=../.install
b install
tree ../.install
```
The output can be seen in the following. The `README.md` files of the dependencies are not installed.
```
../.install
├── bin
│   └── hello
├── include
│   ├── libformat
│   │   ├── export.hxx
│   │   ├── format.hxx
│   │   └── version.hxx
│   ├── libhello
│   │   ├── export.hxx
│   │   ├── hello.hxx
│   │   └── version.hxx
│   └── libprint
│       ├── export.hxx
│       ├── print.hxx
│       └── version.hxx
├── lib
│   ├── libformat-1.0.so
│   ├── libformat.so -> libformat-1.0.so
│   ├── libhello-1.1.so
│   ├── libhello.so -> libhello-1.1.so
│   ├── libprint-1.0.so
│   ├── libprint.so -> libprint-1.0.so
│   └── pkgconfig
│       ├── libformat.shared.pc
│       ├── libhello.shared.pc
│       └── libprint.shared.pc
└── share
    └── doc
        ├── hello
        │   ├── manifest
        │   └── README.md
        ├── libformat
        │   └── manifest
        ├── libhello
        │   └── manifest
        └── libprint
            └── manifest
```
Furthermore, for other packages on `build2-packaging` that build libraries with dependencies not even the `manifest` files of their dependencies get installed. Could this be a bug?

**Comment by boris-kolpackov** [comment]
This is by design. Specifically, the build system's `install` operation only installs targets that are required by the current project (more exactly, by the targets being installed). The canonical example is an executable that depends on a library which in turn can be built shared and static. If the executable links to the shared library then if you install the executable, then only the shared library will get installed. While this is not always the desired behavior, often it is, and this is the only level (i.e., build system) it can be sensibly provided.

For cases where this is not the desired behavior and you would like to install the dependencies completely, you use the package manages (`bpkg`) to install recursively (see the `--recursive` and `--immediate` options): https://build2.org/bpkg/doc/bpkg-pkg-install.xhtml

For a good use-case where both of these methods are used, see: https://build2.org/bpkg/doc/bpkg-pkg-bindist.xhtml

One part that is missing currently is the `install`/`uninstall` commands in `bdep`. I've added an issue for that: https://github.com/build2/build2/issues/320

(As an aside, you could make the `manifest` and other documentation to be installed as part of the library if you list them as prerequisites of the library. But this is not canonical and can get quite repetitive if you have multiple libraries. But I also agree that installing, say, a shared library without any information about its origin or its license is not ideal. Perhaps we could add some "special case" and always install at least the `manifest` if any file from a project is installed.)

**Comment by boris-kolpackov** [comment]
> Do you mean with special case another feature for `b install` to also install the `manifest`?

Yes, basically have special handling of `manifest` in that if any target from a project is installed, then `manifest` is automatically installed as well.

> Or do you rather mean that I should try to add `manifest` to the dependencies of a library, like in the following?

You could but it's quite burdensome to list these prerequisites for every relevant target.

**Comment by lyrahgames** [comment]
> Yes, basically have special handling of `manifest` in that if any target from a project is installed, then `manifest` is automatically installed as well.

I was thinking again about the handling of license copies and the copyright often stated in a README.
Isn't it always necessary to additionally provide these files for any installation/distribution to properly handle any legal cases?
If this is true, then also these would need to be handled like the `manifest`.

**Comment by boris-kolpackov** [comment]
> Isn't it always necessary to additionally provide these files for any installation/distribution to properly handle any legal cases?

Yes, that's true. I've been thinking on this in the background and perhaps we could implement some "special handling" in that we always install any `doc{}` or `legal{}` prerequisites of the `./` target in the root `buildfile`. A bit hackish but probably not too bad.

**Comment by lyrahgames** [comment]
> Yes, that's true. I've been thinking on this in the background and perhaps we could implement some "special handling" in that we always install any `doc{}` or `legal{}` prerequisites of the `./` target in the root `buildfile`. A bit hackish but probably not too bad.

What would you think of introducing a notion of default prerequisites for any target inside a package?
In the `root.build`, it could be specified by initial package creation that every target has `manifest`, `doc{README.md}`, and any `legal{*}` as dependency. Maybe something like the following.

```
# root.build
# ...
target{*}: $src_root/manifest $src_root/doc{README.md} $src_root/legal{**}
# ...
```

I know that this is more complex. But it would fit into the typical way `build2` handles prerequisites and default snippets and, consequently, could be more elegant and even give the user the opportunity to remove the default targets if not needed.

**Comment by boris-kolpackov** [comment]
> What would you think of introducing a notion of default prerequisites for any target inside a package?

I think this could be too unpredictable since this will add prerequisites to every target and for every operation. You are listing these files because you thinking of executables/libraries and `install` but some rule (including ad hoc) for some other operation may see them and decide to do something differently. For example, there may be a rule that converts one or more Markdown files to HTML. And now that it sees `README.md` (and perhaps `LICENSE.md`), it will convert them as well.

**Comment by boris-kolpackov** [comment]
Yes, it's a good point regarding `*-file` manifest values, we may end up installing a manifest that references other files nowhere to be found.

Also, agree about implicit/magic behavior: would be nice to have a general model that covers this.
